### PR TITLE
plug cell map

### PR DESCRIPTION
**Pull request by nicolasaunai**
plug the cellmap to phare

### CODE REVIEW

**Comment by nicolasaunai** [comment]
- [ ] pack_
- [ ] unpack_
- [ ] copy_
- [ ] copyWithTransform_
- [ ] countParticlesIn (can be just the return of cellmap.size in a box

File: src/core/utilities/box/box.h [hunk 1]
```diff
@@ -2,119 +2,280 @@
...
 
-        using type = Type;
-    };
-
-    template<typename T, std::size_t s>
-    Box(Point<T, s> lower, Point<T, s> upper)->Box<T, s>;
+    using iterator = box_iterator<Type, dim>;
+    auto begin() { return iterator{this, lower}; }
 
+    //   // since the 1D scan of the multidimensional box is done assuming C ordering
```
Comment (nicolasaunai) [inline src/core/utilities/box/box.h:88]: outdated comment

File: src/core/numerics/boundary_condition/boundary_condition.h [hunk 2]
```diff
@@ -24,14 +23,15 @@ namespace core
         ParticleIterator applyOutgoingParticleBC(ParticleIterator begin, ParticleIterator end)
         {
             // TODO loop while last partition not equel to begin.
-            auto partitions = partitionner(begin, end, boundaryBoxes_);
+            //            auto partitions = partitionner(begin, end, boundaryBoxes_);
```
Comment (nicolasaunai) [inline src/core/numerics/boundary_condition/boundary_condition.h:26]: TBD if this should go or not

File: src/core/utilities/box/box.h [hunk 3]
```diff
@@ -2,119 +2,280 @@
...
+            return 6;
+    }
 
-    /** this overload of isIn takes a Point and a Container of boxes
-     * and returns true if the Point is at least in one of the boxes.
-     * Returns occurs at the first box the point is in.
-     */
-    template<typename Point, typename BoxContainer, is_iterable<BoxContainer> = dummy::value>
-    bool isIn(Point const& point, BoxContainer const& boxes)
+    auto remove_inside(Box const& to_remove) const
```
Comment (nicolasaunai) [inline src/core/utilities/box/box.h:143]: unused?

File: src/core/utilities/box/box.h [hunk 4]
```diff
@@ -2,119 +2,280 @@
...
+            increment(idim - 1);
+            if (index_[idim - 1] <= box_->upper[idim - 1])
+                index_[idim] = box_->lower[idim];
         }
+    }
 
-        return false;
+    box_iterator operator++()
+    {
+        //   if constexpr (dim == 2)
```
Comment (nicolasaunai) [inline src/core/utilities/box/box.h:227]: take out

File: src/core/utilities/bucketlist.h [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    auto begin() const { return iterator{*this}; }
+
+    auto end();
+
+    auto end() const;
+
+    void empty()
+    {
+        if (capacity() > 3 * size())
```
Comment (nicolasaunai) [inline src/core/utilities/bucketlist.h:55]: magic?

File: src/core/utilities/cellmap.h [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        {
+            auto blist = list_at(cell);
+            if (blist)
+                s += blist->get().size();
+        }
+        return s;
+    }
+
+
+    auto select(box_t const& box) const
```
Comment (nicolasaunai) [inline src/core/utilities/cellmap.h:94]: torm?

File: src/core/utilities/bucketlist.h (hunk 5 above)
Comment (PhilipDeegan) [reply src/core/utilities/bucketlist.h:55]: :mage_man:

File: src/amr/data/particles/particles_data.h [hunk 7]
```diff
@@ -417,96 +368,145 @@ namespace amr
...
+        void copy_(SAMRAI::hier::Box const& overlapBox, ParticlesData const& sourceData)
+        {
+            auto myDomainBox         = this->getBox();
+            auto& srcDomainParticles = sourceData.domainParticles;
 
+            PHARE_LOG_START("ParticleData::copy_ DomainToDomain");
+
+            // first copy particles that fall into our domain array
+            // they can come from the source domain or patch ghost
+            auto destBox  = myDomainBox * overlapBox;
```
Comment (PhilipDeegan) [inline src/amr/data/particles/particles_data.h:380]: can be const?

File: src/amr/data/particles/particles_data.h [hunk 8]
```diff
@@ -417,96 +368,145 @@ namespace amr
...
+            auto new_size = domainParticles.size();
+
+            if (!destBox.empty())
+            {
+                auto destBox_p = toPHAREBox<dim>(destBox);
```
Comment (PhilipDeegan) [inline src/amr/data/particles/particles_data.h:385]: toPHAREBox was removed in #629

Comment (PhilipDeegan) [reply src/amr/data/particles/particles_data.h:385]: https://gist.githubusercontent.com/PhilipDeegan/017af0ae9fe254a8967eb783243998a3/raw/ac908b7cc3ac871017972220a78182edf7a1e145/gistfile1.txt :eyes:

File: src/core/utilities/box/box.h [hunk 9]
```diff
@@ -2,119 +2,280 @@
...
+struct Box
 {
-    /** Represents a 1D, 2D or 3D box of integer or floating point
-     * points.
-     */
-    template<typename Type, std::size_t dim>
-    struct Box
+    Point<Type, dim> lower;
+    Point<Type, dim> upper;
+    std::array<Type, dim> shape;
```
Comment (PhilipDeegan) [inline src/core/utilities/box/box.h:29]: this needs to be kept in sync with the lower upper if they ever change, probably better to just use the shape() that was merged in #629

**Pull request closed without merging**